src/rust/engine/fs/store/src/local.rs

```diff
@@ -283,18 +325,21 @@ impl UnderlyingByteStore for ShardedFSDB {
     expected_digest: Digest,
     src: PathBuf,
   ) -> Result<(), String> {
-    let dest = self.get_tempfile(expected_digest.hash).await?;
     let mut attempts = 0;
     loop {
-      let (mut reader, mut writer) = try_join(tokio::fs::File::open(src.clone()), dest.open())
+      let reader = tokio::fs::File::open(src.clone())
         .await
-        .map_err(|e| e.to_string())?;
-      // TODO: Consider using `fclonefileat` on macOS, which would skip actual copying (read+write), and
-      // instead just require verifying the resulting content after the syscall (read only).
-      let should_retry =
-        !async_verified_copy(expected_digest, src_is_immutable, &mut reader, &mut writer)
-          .await
-          .map_err(|e| e.to_string())?;
+        .map_err(|e| format!("Failed to open {src:?}: {e}"))?;
+
+      // TODO: Consider using `fclonefileat` on macOS or checking for same filesystem+rename on Linux,
+      // which would skip actual copying (read+write), and instead just require verifying the
+      // resulting content after the syscall (read only).
+      let should_retry = self
+        .write_using(expected_digest.hash, |file| {
+          Self::verified_copier(file, expected_digest, src_is_immutable, reader)
+        })
+        .await
+        .is_err();
 
       if should_retry {
```

Comment:
Right now the actual underlying error from the copy will be lost after 10 attempts: should preserve it.

But also, rather than converting from `Err` to bool here, it might be good to actually have a specific return type for "input changed", to avoid retrying things that will never succeed (and to make the "Input changed while reading" debug message always actually accurate).

Reply:
Yeah I'm between a rock and a hard place here, since we don't want to poison the `OnceCell` with code that completed, but failed. Only way I can tell to accomplish that is through the error channel. Perhaps an error type being a union?

Reply:
Yea.